### DISCUSSION TITLE
MIDI_CC volume control for single TG

### DISCUSSION THREAD

**Discussion by Picard-Synth** [probonopd/MiniDexed#689] [Ideas]
New user here - Hi everybody!

I dream of using my 8 sliders of the MIDI keybord as volume sliders of the 8 TGs.
As I understand this would be possible if each TG had a separate MIDI channel. Unfortunately (for this use case), the performance files use always the same channel for all TGs. That means I would need to edit each and every file...

Would it be possible to define 8 MIDI_CC numbers for the TG volumes?

like in mididevice.cpp:
#define MIDI_CC_VOLUME_TG1     81
#define MIDI_CC_VOLUME_TG2     82
#define MIDI_CC_VOLUME_TG3     83
#define MIDI_CC_VOLUME_TG4     84
#define MIDI_CC_VOLUME_TG5     85
#define MIDI_CC_VOLUME_TG6     86
#define MIDI_CC_VOLUME_TG7     87
#define MIDI_CC_VOLUME_TG8     88

Then everything would be easy peasy and working with "stock" performances.
Or are these numbers reserved for future use?

**Comment by diyelectromusic** [comment]
I'm nervous of hard-coding in any specific non-DX7 MIDI mappings into the system, but that doesn't mean we couldn't include a way to define custom mappings via minidexed.ini, performance.ini or some other means.

We have talked from time to time about pulling out the MIDI channel from the performance.ini files into some kind of system/MIDI ini file and this to me seems like a similar thing.

I'll have a think.

Another possibility I've experimented with is exposing the internal "command" interface over MIDI via SysEx, then as well as access to things like "change volume for this TG" it also opens up all sorts of possibilities for custom interfaces to MiniDexed (I used it in my "dual" setup for example).  Basically any menu option would be SysEx-able...

Kevin

**Reply by Picard-Synth** [reply]
I see hard-coding non-DX7 mappings isn't ideal. Also, in case it would de done I see myself longing for the next parameter like panning and so on...
So I agree exposing the interface via SysEx would be the royal road.
What are the chances of achieving something like this?

**Reply by diyelectromusic** [reply]
I had a prototype, so once the latest change is complete, I could take another look.

But it would require a controller being able to be programmed with a custom SysEx message, which might limit the options a little.

Kevin

**Reply by Picard-Synth** [reply]
I use an Edirol PCR-800 which does sysex pretty nicely.
Another option would be the way PreenFM does. It's extensively MIDI-programmable and uses CC for general parameters and NRPN for the vast of operator parameters. Maybe that's a good compromise and easier available for the bulk of controllers (...and users).
[https://ixox.fr/preenfm2/preenfm/midi/](https://ixox.fr/preenfm2/preenfm/midi/)
I'm certainly willing to contribute, if at all possible. I'm good with hardware and/or writing manuals.

**Comment by probonopd** [comment]
> the performance files use always the same channel for all TGs. That means I would need to edit each and every file

Shouldn't be hard to do at all with just a tiny bit of scripting:

```
for file in performance/*; do sed -i 's/\(MIDIChannel\([0-9]*\)\)=.*/\1=\2/' "$file"; done
```

Let me know if I should post a modified set of performances in which every TG is on its own MIDI channel.

**Reply by Picard-Synth** [reply]
Thank you so much for support. After trying a bit I realized that spreading the TGs over the MIDI channels means that I always play only one at once... I think it was a misconception of mine

**Reply by probonopd** [reply]
If your controller can send notes to OMNI, then all MIDI notes will receive them no matter which channel they are set to.

**Reply by Picard-Synth** [reply]
Are there actually controllers that can send notes to omni? Wouldn't this mean sending 16 note on and 16 note off messages per key press? What a mess... ;-) I'm not aware of such a scheme to be part of the MIDI specification.

**Reply by probonopd** [reply]
You are right, I was thinking in terms of DAWs where you can send things to OMNI. Indeed in the MIDI spec sending to OMNI is not possible and you'd have to send one message per channel, which is inefficient. (It shows that I never used MIDI "back in the day" before sound was created with plugins in the DAW.)
